File: .github/scripts/setup_routeros.py
import routeros_api
import os


def main():
    user = os.environ.get("ROS_USERNAME")
    pswd = os.environ.get("ROS_PASSWORD")
    ip_addr = os.environ.get("ROS_IP_ADDRESS")

    connection = routeros_api.RouterOsApiPool(
        ip_addr, username=user, password=pswd, port=8728, plaintext_login=True)
    api = connection.get_api()

    # Set up bridge

    bridge = api.get_resource("/interface/bridge")
    bridge.add(name="bridge")

    # Set up certificates to allow use of REST API

    certificate = api.get_resource("/certificate")
    certificate.add(name="root-cert", common_name="MyRouter",
                    days_valid="3650", key_usage="key-cert-sign,crl-sign")
    certificate.add(name="https-cert",
                    common_name="MyRouter", days_valid="3650")
    certs = certificate.get()
    root_cert_id = [x['id'] for x in certs if x['name'] == "root-cert"][0]
    http_cert_id = [x['id'] for x in certs if x['name'] == "https-cert"][0]
    api.get_binary_resource("/").call("certificate/sign",
                                      {"id": bytes(root_cert_id, "utf-8")})
    api.get_binary_resource("/").call("certificate/sign",
                                      {"id": bytes(http_cert_id, "utf-8"), "ca": b"root-cert"})
    services = api.get_resource("/ip/service")
    for x in services.get():
        if x['name'] in ['www-ssl', 'api-ssl']:
            services.set(id=x['id'],certificate="https-cert", disabled="false")

    # Create a DHCP pool

    pools = api.get_resource("/ip/pool")
    pools.add(name="dhcp", ranges="192.168.88.100-192.168.88.200")

    wireguard = api.get_resource("/interface/wireguard")
    wireguard.add(name="wg1")

    # Output the list of interfaces

    print(api.get_resource("/interface").get())


if __name__ == "__main__":
    main()

File: .github/scripts/setup_routeros.go
package main

import (
	"log"
	"os"
	"strings"
	"time"

	"github.com/go-routeros/routeros"
)

var (
	commands = []string{
		"/certificate/add name=ssl common-name=router key-size=prime256v1",
		"/certificate/sign .id=ssl",
		"/ip/service/set .id=www-ssl disabled=no certificate=ssl",
		"/ip/service/set .id=api-ssl disabled=no certificate=ssl",
		"/interface/bridge/add name=bridge",
		"/ip/pool/add name=dhcp ranges=192.168.88.100-192.168.88.200",
		"/interface/wireguard/add name=wg1",
		"/interface/list/add name=list",
	}
)

func main() {
	username := os.Getenv("ROS_USERNAME")
	password := os.Getenv("ROS_PASSWORD")
	host := os.Getenv("ROS_IP_ADDRESS")

	var err error
	var client *routeros.Client
	for i := 0; i < 12; i++ {
		log.Printf("Connection attempt #%v... ", i)
		client, err = routeros.Dial(host+":8728", username, password)
		if err == nil {
			break
		}
		log.Println(err)
		time.Sleep(5 * time.Second)
	}
	if err != nil {
		log.Fatal("Host is not available")
	}
	defer client.Close()

	for _, command := range commands {
		cmd := strings.Split(command, " ")

		for i, c := range cmd {
			if strings.ContainsRune(c, '=') && len(c) > 0 && c[0] != '=' {
				cmd[i] = "=" + c
			}
		}

		log.Println(cmd)

		res, err := client.RunArgs(cmd)
		if err != nil {
			log.Fatal(err)
		}

		log.Println(res)
	}
}
